### PR TITLE
Feature/dbiterator

### PR DESCRIPTION
**Pull request by Mr-Leshiy**

### CODE REVIEW

File: include/veriblock/storage/block_iterator.hpp [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+  virtual void next() = 0;
+
+  virtual BlockIndex<BlockT> value() const = 0;
+
+  virtual hash_t key() const = 0;
+
+  virtual bool valid() const = 0;
+
+  virtual bool seek_start() = 0;
```
Comment (Warchant) [inline include/veriblock/storage/block_iterator.hpp:42]: please write docs for these funcs.
can seek_start fail? if not, why does it return `bool`?

File: include/veriblock/storage/block_provider.hpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ * An abstraction over on-disk storage iterator.
+ *
+ * veriblock-pop-cpp does not dictate how to iterate over the storage. Altchain
+ * must create derived class.
+ *
+ * @ingroup interfaces
+ */
+
+template <typename BlockT>
+struct BlockProvider {
```
Comment (Warchant) [inline include/veriblock/storage/block_provider.hpp:32]: I think this interface should be hidden in namespace `details` and then there would be another interface which implements this interface specific for each block, e.g.
```
struct BlockProvider : public GenericBlockProvider<AltBlock>, public GenericBlockProvider<VbkBlock>, public GenericBlockProvider<BtcBlock> {}
```

File: src/storage/util.cpp [hunk 3]
```diff
@@ -2,11 +2,64 @@
...
+
+  std::vector<index_t> blocks;
+
+  for (; it->valid(); it->next()) {
+    blocks.push_back(it->value());
+  }
+
+  hash_t tip_hash;
+  if (!provider.getTipHash(tip_hash)) {
+    return state.Invalid("bad-provider", "cannot get tip");
```
Comment (Warchant) [inline src/storage/util.cpp:33]:
```suggestion
    return state.Invalid("bad-tip", fmt::format("Can not read tip of {} block tip", block_t::name()));
```

File: src/storage/util.cpp [hunk 4]
```diff
@@ -2,11 +2,64 @@
...
+  auto* t = out.getBestChain().tip();
+  VBK_ASSERT(t != nullptr);
+
+  return true;
+}
+
+bool LoadAllTrees(AltBlockTree& tree,
+                  BlockProvider<BtcBlock>& btc_provider,
+                  BlockProvider<VbkBlock>& vbk_provider,
+                  BlockProvider<AltBlock>& alt_provider,
```
Comment (Warchant) [inline src/storage/util.cpp:49]: pass single provider here, then it will be just casted to specific types when passed by ref

**Review by Warchant — CHANGES_REQUESTED** [review]

File: src/storage/util.cpp (hunk 3 above)
Comment (Mr-Leshiy) [reply src/storage/util.cpp:33]: fixed

File: src/storage/util.cpp [hunk 5]
```diff
@@ -2,11 +2,65 @@
...
+  using index_t = typename BlockTreeT::index_t;
+  using hash_t = typename index_t::hash_t;
+
+  std::vector<index_t> blocks;
+
+  auto it = provider.getBlockIterator();
+  for (it->seek_start(); it->valid(); it->next()) {
+    index_t val;
+    if (!it->value(val)) {
+      return state.Invalid("bad-it", "Can not get block value");
```
Comment (Warchant) [inline src/storage/util.cpp:25]:
```suggestion
      return state.Invalid("bad-value", "Can not get block value");
```

File: src/storage/util.cpp [hunk 6]
```diff
@@ -2,11 +2,65 @@
...
+    }
+    blocks.push_back(val);
+  }
+
+  hash_t tip_hash;
+  if (!provider.getTipHash(tip_hash)) {
+    return state.Invalid("bad-tip",
```
Comment (Warchant) [inline src/storage/util.cpp:32]:
```suggestion
    return state.Invalid(index_t::block_t::name() + "-bad-tip",
```

Comment (Warchant) [reply src/storage/util.cpp:32]: do same for "bad-value"

**Review by Warchant — APPROVED** [review]

File: src/storage/util.cpp (hunk 6 above)
Comment (Mr-Leshiy) [reply src/storage/util.cpp]: fixed

File: src/storage/util.cpp (hunk 5 above)
Comment (Mr-Leshiy) [reply src/storage/util.cpp]: fixed

**Pull request merged**